src/service.py

```diff
@@ -0,0 +1,165 @@
+
+        Raises:
+            ServiceRestartError: if the installation of the snap failed.
+            PackageInstallError: if the package installation failed.
+        """
+        try:
+            # Add ppa that hosts the jenkins-agent package
+            repositories = apt.RepositoryMapping()
+            if "deb-ppa.launchpadcontent.net/tphan025/ppa-jammy" not in repositories:
+                repositories.add(
+                    apt.DebianRepository(
+                        enabled=True,
+                        repotype="deb",
+                        uri="https://ppa.launchpadcontent.net/tphan025/ppa/ubuntu/",
+                        # TODO: depends on the series of the charm unit
+                        # set the release accordingly
+                        release="jammy",
+                        groups=["main"],
+                    )
+                )
+                apt.import_key(PPA_GPG_KEY_ID)
+            # Install the apt package
+            apt.update()
+            # Dependencies
+            apt.add_package("openjdk-17-jre")
+            # Jenkins agent package from PPA
+            apt.add_package(AGENT_PACKAGE_NAME)
+            systemd.service_restart(AGENT_SERVICE_NAME)
+        except systemd.SystemdError as exc:
+            raise ServiceRestartError("Error starting the agent service") from exc
+        except (apt.PackageError, apt.PackageNotFoundError) as exc:
+            raise PackageInstallError("Error installing the agent package") from exc
+
+    def restart(self) -> None:
+        """Start the agent service.
+
+        Raises:
+            ServiceRestartError: when restarting the service fails
+        """
+        # Render template and write to appropriate file if only credentials are set
+        if credentials := self.state.agent_relation_credentials:
+            with open("templates/jenkins_agent_env.conf.j2", "r", encoding="utf-8") as file:
+                template = Template(file.read())
+            # fetch credentials and set them as environments
+            environments = {
+                "JENKINS_TOKEN": credentials.secret,
+                "JENKINS_URL": credentials.address,
+                "JENKINS_AGENT": self.state.agent_meta.name,
+            }
+            # render template
+            rendered = template.render(environments=environments)
+            # Ensure that service conf directory exist
+            config_dir = Path(SYSTEMD_SERVICE_CONF_DIR)
+            config_dir.mkdir(parents=True, exist_ok=True)
+            # Write the conf file
+            logger.info("Rendering agent configuration")
+            logger.debug("%s", environments)
+            # file name (override.conf) is important for the service to import envvars
+            self._render_file(f"{config_dir.resolve().as_posix()}/override.conf", rendered, 0o644)
+        try:
+            systemd.service_restart(AGENT_SERVICE_NAME)
+        except systemd.SystemdError as exc:
```

Comment:
Should file system interaction errors also be handled here? Any other errors that could occur?

Reply:
That's right, there's a `SystemError` that is raised if the subprocess call of `systemctl` fails, I added handling for that case